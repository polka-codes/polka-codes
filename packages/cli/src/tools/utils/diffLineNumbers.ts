// generated by polka.codes
// Utility functions for parsing and annotating unified diff format with line numbers

export interface DiffHunk {
  oldStart: number
  oldCount: number
  newStart: number
  newCount: number
  header: string
}

/**
 * Parse a unified diff hunk header
 * Format: @@ -oldStart,oldCount +newStart,newCount @@ optional context
 */
export function parseHunkHeader(header: string): DiffHunk | null {
  const match = header.match(/^@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@/)
  if (!match) return null

  return {
    oldStart: parseInt(match[1], 10),
    oldCount: match[2] ? parseInt(match[2], 10) : 1,
    newStart: parseInt(match[3], 10),
    newCount: match[4] ? parseInt(match[4], 10) : 1,
    header,
  }
}

/**
 * Annotate a diff with line numbers for additions and deletions
 * Adds [Line N] for additions and [Line N removed] for deletions
 */
export function annotateDiffWithLineNumbers(diff: string): string {
  const lines = diff.split('\n')
  const annotatedLines: string[] = []
  let currentNewLine = 0
  let currentOldLine = 0
  let inHunk = false

  for (const line of lines) {
    // Check if this is a hunk header
    if (line.startsWith('@@')) {
      const hunk = parseHunkHeader(line)
      if (hunk) {
        currentOldLine = hunk.oldStart
        currentNewLine = hunk.newStart
        inHunk = true
      }
      annotatedLines.push(line)
      continue
    }

    // Skip file headers and other metadata
    if (line.startsWith('diff --git') || line.startsWith('index ') || line.startsWith('---') || line.startsWith('+++')) {
      annotatedLines.push(line)
      inHunk = false
      continue
    }

    // Only annotate lines within hunks
    if (!inHunk) {
      annotatedLines.push(line)
      continue
    }

    // Handle diff lines
    if (line.startsWith('+') && !line.startsWith('+++')) {
      // Addition - annotate with new file line number
      annotatedLines.push(`${line} [Line ${currentNewLine}]`)
      currentNewLine++
    } else if (line.startsWith('-') && !line.startsWith('---')) {
      // Deletion - annotate with old file line number
      annotatedLines.push(`${line} [Line ${currentOldLine} removed]`)
      currentOldLine++
    } else if (line.startsWith(' ')) {
      // Context line - both line numbers advance
      annotatedLines.push(line)
      currentOldLine++
      currentNewLine++
    } else {
      // Other lines (shouldn't happen in a proper diff, but handle gracefully)
      annotatedLines.push(line)
    }
  }

  return annotatedLines.join('\n')
}

/**
 * Extract line ranges from a diff for a specific file
 * Returns an array of line ranges that were modified
 */
export function extractModifiedLineRanges(
  diff: string,
  filePath?: string,
): Array<{ start: number; end: number; type: 'added' | 'removed' | 'modified' }> {
  const lines = diff.split('\n')
  const ranges: Array<{ start: number; end: number; type: 'added' | 'removed' | 'modified' }> = []
  let currentFile = ''
  let currentNewLine = 0
  let currentOldLine = 0
  let inHunk = false
  let addedLines: number[] = []
  let removedLines: number[] = []

  for (const line of lines) {
    // Track current file
    if (line.startsWith('diff --git')) {
      // Extract filename from diff header
      const match = line.match(/diff --git a\/(.*?) b\//)
      if (match) {
        currentFile = match[1]
      }
      // Reset for new file
      addedLines = []
      removedLines = []
      continue
    }

    // Skip if we're looking for a specific file and this isn't it
    if (filePath && currentFile !== filePath) {
      continue
    }

    // Check if this is a hunk header
    if (line.startsWith('@@')) {
      // Process any accumulated changes from previous hunk
      if (addedLines.length > 0 || removedLines.length > 0) {
        processAccumulatedChanges(addedLines, removedLines, ranges)
        addedLines = []
        removedLines = []
      }

      const hunk = parseHunkHeader(line)
      if (hunk) {
        currentOldLine = hunk.oldStart
        currentNewLine = hunk.newStart
        inHunk = true
      }
      continue
    }

    if (!inHunk) continue

    // Track modifications
    if (line.startsWith('+') && !line.startsWith('+++')) {
      addedLines.push(currentNewLine)
      currentNewLine++
    } else if (line.startsWith('-') && !line.startsWith('---')) {
      removedLines.push(currentOldLine)
      currentOldLine++
    } else if (line.startsWith(' ')) {
      // Context line - process any accumulated changes
      if (addedLines.length > 0 || removedLines.length > 0) {
        processAccumulatedChanges(addedLines, removedLines, ranges)
        addedLines = []
        removedLines = []
      }
      currentOldLine++
      currentNewLine++
    }
  }

  // Process any remaining accumulated changes
  if (addedLines.length > 0 || removedLines.length > 0) {
    processAccumulatedChanges(addedLines, removedLines, ranges)
  }

  return ranges
}

function processAccumulatedChanges(
  addedLines: number[],
  removedLines: number[],
  ranges: Array<{ start: number; end: number; type: 'added' | 'removed' | 'modified' }>,
) {
  if (addedLines.length > 0 && removedLines.length > 0) {
    // Modified lines
    if (addedLines.length > 0) {
      ranges.push({
        start: Math.min(...addedLines),
        end: Math.max(...addedLines),
        type: 'modified',
      })
    }
  } else if (addedLines.length > 0) {
    // Only additions
    ranges.push({
      start: Math.min(...addedLines),
      end: Math.max(...addedLines),
      type: 'added',
    })
  } else if (removedLines.length > 0) {
    // Only removals
    ranges.push({
      start: Math.min(...removedLines),
      end: Math.max(...removedLines),
      type: 'removed',
    })
  }
}
