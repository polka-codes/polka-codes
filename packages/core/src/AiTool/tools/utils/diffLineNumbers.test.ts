// generated by polka.codes
import { describe, expect, it } from 'bun:test'
import { annotateDiffWithLineNumbers, extractModifiedLineRanges, parseHunkHeader } from './diffLineNumbers'

describe('diffLineNumbers', () => {
  describe('parseHunkHeader', () => {
    it('should parse standard hunk header', () => {
      const header = '@@ -10,5 +12,7 @@ function example() {'
      const result = parseHunkHeader(header)
      expect(result).toEqual({
        oldStart: 10,
        oldCount: 5,
        newStart: 12,
        newCount: 7,
        header,
      })
    })

    it('should parse hunk header without counts', () => {
      const header = '@@ -1 +1 @@'
      const result = parseHunkHeader(header)
      expect(result).toEqual({
        oldStart: 1,
        oldCount: 1,
        newStart: 1,
        newCount: 1,
        header,
      })
    })

    it('should return null for invalid header', () => {
      const result = parseHunkHeader('not a hunk header')
      expect(result).toBeNull()
    })
  })

  describe('annotateDiffWithLineNumbers', () => {
    it('should annotate additions and deletions with line numbers', () => {
      const diff = `diff --git a/test.js b/test.js
index abc123..def456 100644
--- a/test.js
+++ b/test.js
@@ -10,5 +10,6 @@ function example() {
   console.log('context line 1');
-  console.log('removed line');
+  console.log('added line 1');
+  console.log('added line 2');
   console.log('context line 2');`

      const annotated = annotateDiffWithLineNumbers(diff)
      expect(annotated).toContain('[Line 11 removed]')
      expect(annotated).toContain('[Line 11]')
      expect(annotated).toContain('[Line 12]')
    })

    it('should handle multiple hunks', () => {
      const diff = `@@ -1,3 +1,3 @@
 context
-old
+new
@@ -10,2 +10,3 @@
 more context
+addition`

      const annotated = annotateDiffWithLineNumbers(diff)
      expect(annotated).toMatchSnapshot()
    })
  })

  describe('extractModifiedLineRanges', () => {
    it('should extract line ranges for additions', () => {
      const diff = `@@ -10,3 +10,5 @@
 context
+added line 1
+added line 2
 more context`

      const ranges = extractModifiedLineRanges(diff)
      expect(ranges).toEqual([
        {
          start: 11,
          end: 12,
          type: 'added',
        },
      ])
    })

    it('should extract line ranges for deletions', () => {
      const diff = `@@ -10,5 +10,3 @@
 context
-removed line 1
-removed line 2
 more context`

      const ranges = extractModifiedLineRanges(diff)
      expect(ranges).toEqual([
        {
          start: 11,
          end: 12,
          type: 'removed',
        },
      ])
    })

    it('should extract line ranges for modifications', () => {
      const diff = `@@ -10,3 +10,3 @@
 context
-old line
+new line
 more context`

      const ranges = extractModifiedLineRanges(diff)
      expect(ranges).toEqual([
        {
          start: 11,
          end: 11,
          type: 'modified',
        },
      ])
    })

    it('should handle multiple changes in one hunk', () => {
      const diff = `@@ -10,7 +10,8 @@
 context
+addition 1
 unchanged
-deletion
+replacement
 unchanged
+addition 2`

      const ranges = extractModifiedLineRanges(diff)
      expect(ranges).toMatchSnapshot()
    })
  })
})
