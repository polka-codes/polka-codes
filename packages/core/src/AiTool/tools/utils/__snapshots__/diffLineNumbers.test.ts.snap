// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`diffLineNumbers annotateDiffWithLineNumbers should handle multiple hunks 1`] = `
"@@ -1,3 +1,3 @@
 context
-old [Line 2 removed]
+new [Line 2]
@@ -10,2 +10,3 @@
 more context
+addition [Line 11]"
`;

exports[`diffLineNumbers extractModifiedLineRanges should handle multiple changes in one hunk 1`] = `
[
  {
    "end": 11,
    "start": 11,
    "type": "added",
  },
  {
    "end": 13,
    "start": 13,
    "type": "modified",
  },
  {
    "end": 15,
    "start": 15,
    "type": "added",
  },
]
`;
